File: openmole/core/org.openmole.core.commons.fileservice/src/main/java/org/openmole/core/commons/fileservice/internal/FileService.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.commons.fileservice.internal;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.core.commons.fileservice.IFileService;
import org.openmole.commons.tools.cache.AssociativeCache;
import org.openmole.commons.tools.cache.ICachable;
import org.openmole.commons.tools.filecache.FileCacheDeleteOnFinalize;
import org.openmole.commons.tools.filecache.IFileCache;
import org.openmole.commons.tools.io.IHash;
import org.openmole.commons.tools.io.TarArchiver;

public class FileService implements IFileService {

    AssociativeCache<String, IHash> hashCach = new AssociativeCache<String, IHash>(AssociativeCache.SOFT, AssociativeCache.SOFT);
    AssociativeCache<String, IFileCache> archiveCach = new AssociativeCache<String, IFileCache>(AssociativeCache.SOFT, AssociativeCache.SOFT);


    @Override
    public IHash getHashForFile(File file) throws InternalProcessingError, InterruptedException {
       return getHashForFile(file, file);
    }

    @Override
    public File getArchiveForDir(File file) throws InternalProcessingError, InterruptedException {
        return getArchiveForDir(file, file);
    }

    @Override
    public IHash getHashForFile(final File file, Object cacheLength) throws InternalProcessingError, InterruptedException {
        return hashCach.getCache(cacheLength, file.getAbsolutePath(), new ICachable<IHash>() {

            @Override
            public IHash compute() throws InternalProcessingError, InterruptedException {
                try {
                    return Activator.getHashService().computeHash(file);
                } catch (IOException ex) {
                    throw new InternalProcessingError(ex);
                }
            }
        });
    }

    @Override
    public File getArchiveForDir(final File file, Object cacheLenght) throws InternalProcessingError, InterruptedException {
        return archiveCach.getCache(cacheLenght, file.getAbsolutePath(), new ICachable<IFileCache>() {

            @Override
            public IFileCache compute() throws InternalProcessingError, InterruptedException {
                try {
                    File ret = Activator.getWorkspace().newFile("archive", ".tar");
                    OutputStream os = new FileOutputStream(ret);

                    try {
                        new TarArchiver().createDirArchiveWithRelativePath(file, os);
                    } finally {
                        os.close();
                    }
                    
                    return new FileCacheDeleteOnFinalize(ret);
                } catch (IOException ex) {
                    throw new InternalProcessingError(ex);
                }
            }
        }).getFile();
    }
}

File: openmole/misc/org.openmole.misc.fileservice/src/main/scala/org/openmole/misc/fileservice/FileService.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.misc.fileservice

import com.ice.tar.TarOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.logging.Logger
import org.openmole.misc.tools.cache.AssociativeCache
import org.openmole.misc.tools.io.FileUtil._
import org.openmole.misc.tools.io.TarArchiver._
import org.openmole.misc.tools.service.IHash
import org.openmole.misc.filecache.FileCacheDeleteOnFinalize
import org.openmole.misc.filecache.IFileCache
import org.openmole.misc.hashservice.HashService
import org.openmole.misc.updater.Updater
import org.openmole.misc.workspace.ConfigurationLocation
import org.openmole.misc.workspace.Workspace

object FileService {
  val GCInterval = new ConfigurationLocation("FileService", "GCInterval")
  Workspace += (GCInterval, "PT5M")

  class CachedArchiveForDir(file: File, val lastModified: Long) extends FileCacheDeleteOnFinalize(file)

  private[fileservice] val hashCache = new AssociativeCache[String, IHash]
  private[fileservice] val archiveCache = new AssociativeCache[String, CachedArchiveForDir]

  Updater.delay(new FileServiceGC, Workspace.preferenceAsDurationInMs(FileService.GCInterval))

  def hash(file: File): IHash =
    if (file.isDirectory) hash(archiveForDir(file).file(false), file)
    else hash(file, file)

  def archiveForDir(file: File): IFileCache = archiveForDir(file, file)

  def hash(key: Object, file: File): IHash = hashCache.cache(key, file.getAbsolutePath, HashService.computeHash(file))

  def archiveForDir(key: Object, file: File): IFileCache = {

    //invalidateDirCacheIfModified(file, cacheLenght)

    archiveCache.cache(key, file.getAbsolutePath, {
      val ret = Workspace.newFile("archive", ".tar");
      val os = new TarOutputStream(new FileOutputStream(ret))
      try os.createDirArchiveWithRelativePathNoVariableContent(file)
      finally os.close

      new CachedArchiveForDir(ret, file.lastModification)
    })
  }

  private[fileservice] def invalidateDirCacheIfModified(file: File, cacheLenght: Object) =
    archiveCache.cached(cacheLenght, file.getAbsolutePath) match {
      case None ⇒
      case Some(cached) ⇒
        if (cached.lastModified < file.lastModification) {
          archiveCache.invalidateCache(cacheLenght, file.getAbsolutePath)
        }
    }

}
